### PR TITLE
fix(associative-memory): reset an evicted row before a new key writes into it

### LINKED ISSUE #456: Associative memory never resets an evicted row, so a freshly allocated key inherits the evicted key's values, utility, and counts (small-table ablations score below chance)
## Summary

`AssociativeMemoryLearner.update` (`alberta_framework/core/associative_memory.py`) documents that "the table budget is reused by **replacing** low-utility rows" and that "row values are decayed and updated toward the observed next label". In `row_step`, when a key is not found and a slot is reused by eviction, only `keys[slot]` is overwritten: `old_row = carry.values[slot]`, `old_utility = carry.utility[slot]`, and `counts.at[slot].add(1.0)` are carried over from the **previous occupant**. The row therefore no longer represents its key: a freshly allocated key inherits the evicted key's values, utility, and counts, and predicts the previous occupant's label at high confidence until enough of its own writes accumulate.

Because `AlbertaPipeline` uses `associative.predict(...).probabilities` as the Step 2 feature vector feeding Steps 3/4, and `run_step2_associative_smoke` reports `final_window_nll` off the same path, any budget ablation below the default `max_features=512` (which records 0 replacements) measures inheritance, not capacity — a fresh key emits the previous occupant's label at p≈1, so the reported NLL/accuracy for a smaller table is systematically anti-predictive rather than merely degraded.

## Falsifier (deterministic, pristine `main` @ `b8a37de`)

Isolated: `vocab_size=4, block_size=2, suffix_length=2, feature_family="position_token", max_features=2`; write context A with label 3 thirty times, then write context B with label 1 once (evicting A):

```
B seen once, label=1.  found = [1, 1, 0]
  values for B's rows = [[0.0, 1.0, 0.0, 3.995], ...]   <- 3.995 at class 3 (A's label) vs 1.0 at class 1 (B's only write)
  logits argmax = 3
  B's key only starts predicting its own label after 4 total writes
```

Steady state on a periodic 4-class stream (`repro_evict_stream.py`, 320 replacements at small tables):

```
max_features=  2  steady-state acc=0.000  loss=1.406  (chance acc=0.25, uniform loss=1.386)  replacements=318
max_features=  4  steady-state acc=0.000  loss=1.406  replacements=316
max_features=  8  steady-state acc=1.000  loss=0.000  replacements=0
```

Accuracy 0.0 — below chance — because every fresh key emits the evicted key's label. Expected: a slot reused for a new key starts from an empty row, zero utility, and a count of one (with the fix the same stream gives `acc 0.25 / 0.50 / 1.00` for `max_features 2 / 4 / 8`).

## Dedup

Open #446 / PR #449 (deepanshu-yd) harden `AssociativeMemoryConfig` scalar/int validation — a different surface in the same module (`_validate_config`, config class); merge-tree composes cleanly. Adjacent closed: PRs #196, #90 (associative-memory 0×inf / inf−inf guards). `gh issue/pr list --search "eviction" / "replacement row"`: #432 (ExperientialMemoryPolicy, different module), #407/#371 (UPGD/CBP recycling budgets, unrelated). No issue covers the eviction row reset.

Fix PR follows (failing-test-first).

### PR DESCRIPTION
**Pull request by Prabhat1308**
Closes #456.

## Issue

`row_step` overwrote only the key of a slot reused by eviction; the evicted occupant's values, utility, and count were carried over, so a freshly allocated key predicted the previous key's label at high confidence for several writes and every budget ablation below the default table size scored below chance (`acc=0.000` at `max_features=2/4` on a periodic 4-class stream).

## New state

When the key is not found, `row_step` starts from an empty row (`zeros`), zero utility, and a count of `1.0`; found keys keep decaying/accumulating exactly as before. Allocation and replacement counters, `last_update`, and the found-key path are unchanged.

Failing-test-first: `test_evicted_row_is_reset_before_a_new_key_writes_into_it` (30 writes of A@3, one write of B@1 evicting A → B's found rows carry no mass at class 3, argmax is 1, counts are 1) fails on `main` and passes here.

## Evidence

Steady-state stream from #456 at this head:

```text
max_features=  2  steady-state acc=0.250  loss=1.054  replacements=279
max_features=  4  steady-state acc=0.500  loss=0.703  replacements=199
max_features=  8  steady-state acc=1.000  loss=0.000  replacements=0
```

Verification at `a0090c5777f641dc690814e75949b1d989914506`:

```text
.venv/bin/python -m pytest tests/test_associative_memory.py tests/test_step2*.py tests/test_pipeline*.py -q -o addopts=""   -> 54 passed
.venv/bin/python -m ruff check .                                                                                     -> All checks passed!
.venv/bin/python -m mypy                                                                                             -> Success: no issues found in 164 source files
```

Composes cleanly with open #449 (`git merge-tree` clean; #449 edits config validation, this edits the write path). `core/associative_memory.py` is not a registered evidence source; no benchmark lane, seeds, or `outputs/` artifacts were touched; no `CHANGELOG.md` edit (maintainer-recorded).

CLI sessions cannot mint GitHub attachment URLs, so the run output above is inline rather than attached.

- [x] logs: inline above (focused pytest, ruff, mypy, falsifier output at the evidence head)

Provider/model/client: anthropic / claude-fable-5 / Claude Code 2.1.233 (contribute-to-asi skill revision 72e4239e).

### CODE REVIEW

**Review by lalalune — APPROVED** [review]
Independent exact-head review passed at a0090c5777f641dc690814e75949b1d989914506. Evicted-row allocation is equivalent to a fresh slot, including deterministic tie reuse, full row reset, eager/JIT parity, and unaffected-slot preservation. Focused and adjacent tests, Ruff, full mypy, current-main composition, and all 20 hosted checks passed. No registered evidence source is changed.
